bundles/org.openhab.binding.kaleidescape/src/main/java/org/openhab/binding/kaleidescape/internal/handler/KaleidescapeHandler.java

```diff
@@ -0,0 +1,617 @@
+    private long lastEventReceived = 0;
+    private int updatePeriod = 0;
+    private String componentType = EMPTY;
+
+    protected KaleidescapeConnector connector = new KaleidescapeDefaultConnector();
+    protected int metaRuntimeMultiple = 1;
+    protected int volume = 0;
+    protected boolean volumeEnabled = false;
+    protected boolean isMuted = false;
+    protected String friendlyName = EMPTY;
+    protected Object sequenceLock = new Object();
+
+    public KaleidescapeHandler(Thing thing, SerialPortManager serialPortManager, HttpClient httpClient) {
+        super(thing);
+        this.serialPortManager = serialPortManager;
+        this.httpClient = httpClient;
+    }
+
+    protected void updateChannel(String channelUID, State state) {
+        this.updateState(channelUID, state);
+    }
+
+    protected void updateDetailChannel(String channelUID, State state) {
+        this.updateState(DETAIL + channelUID, state);
+    }
+
+    protected void updateThingProperty(String name, String value) {
+        thing.setProperty(name, value);
+    }
+
+    @Override
+    public void initialize() {
+        final String uid = this.getThing().getUID().getAsString();
+        KaleidescapeThingConfiguration config = getConfigAs(KaleidescapeThingConfiguration.class);
+        @Nullable
+        final String componentType = config.componentType;
+
+        // Check configuration settings
+        String configError = null;
+        final String serialPort = config.serialPort;
+        final String host = config.host;
+        final Integer port = config.port;
+        final Integer updatePeriod = config.updatePeriod;
+
+        if ((serialPort == null || serialPort.isEmpty()) && (host == null || host.isEmpty())) {
+            configError = "undefined serialPort and host configuration settings; please set one of them";
+        } else if (host == null || host.isEmpty()) {
+            if (serialPort != null && serialPort.toLowerCase().startsWith("rfc2217")) {
+                configError = "use host and port configuration settings for a serial over IP connection";
+            }
+        } else {
+            if (port == null) {
+                configError = "undefined port configuration setting";
+            } else if (port <= 0) {
+                configError = "invalid port configuration setting";
+            }
+        }
+
+        if (configError != null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, configError);
+            return;
+        }
+
+        if (componentType != null) {
+            this.componentType = componentType;
+        }
+
+        if (updatePeriod != null) {
+            this.updatePeriod = updatePeriod;
+        }
+
+        boolean channelsRemoved = false;
+        List<Channel> channels = new ArrayList<>(this.getThing().getChannels());
+
+        // check if volume is enabled, if not remove the volume & mute channels
+        if (config.volumeEnabled) {
```

Comment:
I wonder if you would create a thing with volum disabled and than edit the thing to enable the volume if the channels will come back? Not sure if that works? You might need to recreate them.

Reply:
I did not realize remove was permanent. I changed it to not remove the channels.